packages/nouns-contracts/contracts/Rewards.sol

```diff
@@ -0,0 +1,280 @@
+ * ░░░░░░██░░░████░░██░░░████░░░ *
+ * ░░██████░░░████████░░░████░░░ *
+ * ░░██░░██░░░████░░██░░░████░░░ *
+ * ░░██░░██░░░████░░██░░░████░░░ *
+ * ░░░░░░█████████░░█████████░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ *********************************/
+
+pragma solidity ^0.8.19;
+
+import { INounsDAOLogicV3 } from './interfaces/INounsDAOLogicV3.sol';
+import { INounsAuctionHouseV2 } from './interfaces/INounsAuctionHouseV2.sol';
+import { NounsDAOV3Types } from './governance/NounsDAOInterfaces.sol';
+import { ERC721 } from '@openzeppelin/contracts-v5/token/ERC721/ERC721.sol';
+import { IERC20 } from '@openzeppelin/contracts-v5/token/ERC20/IERC20.sol';
+import { Ownable } from '@openzeppelin/contracts-v5/access/Ownable.sol';
+
+contract Rewards is ERC721('NounsClientIncentives', 'NounsClientIncentives'), Ownable {
+    INounsDAOLogicV3 public immutable nounsDAO;
+    INounsAuctionHouseV2 public immutable auctionHouse;
+
+    uint32 public nextTokenId = 1;
+
+    uint256 public lastProcessedAuctionId;
+    uint32 public nextProposalIdToReward;
+    uint256 public nextProposalRewardFirstAuctionId;
+    uint256 lastProposalRewardsUpdate = block.timestamp;
+
+    RewardParams public params;
+
+    IERC20 public ethToken; // TODO make configurable
+
+    mapping(uint32 clientId => uint256 balance) public clientBalances;
+
+    struct ClientData {
+        address payoutWallet;
+    }
+
+    struct RewardParams {
+        uint32 minimumRewardPeriod;
+        uint8 numProposalsEnoughForReward;
+        uint16 proposalRewardBps;
+        uint16 votingRewardBps;
+        uint16 auctionRewardBps;
+        uint16 proposalEligibilityQuorumBps;
+    }
+
+    constructor(
+        address owner,
+        address nounsDAO_,
+        address auctionHouse_,
+        address ethToken_,
+        uint32 nextProposalIdToReward_,
+        uint256 lastProcessedAuctionId_,
+        uint256 nextProposalRewardFirstAuctionId_,
+        RewardParams memory rewardParams
+    ) Ownable(owner) {
+        nounsDAO = INounsDAOLogicV3(nounsDAO_);
+        auctionHouse = INounsAuctionHouseV2(auctionHouse_);
+        ethToken = IERC20(ethToken_);
+        nextProposalIdToReward = nextProposalIdToReward_;
+        lastProcessedAuctionId = lastProcessedAuctionId_;
+        nextProposalRewardFirstAuctionId = nextProposalRewardFirstAuctionId_;
+        params = rewardParams;
+    }
+
+    struct Temp {
+        uint256 numEligibleVotes;
+        uint256 rewardPerProposal;
+        uint256 rewardPerVote;
+    }
+
+    function setParams(RewardParams calldata newParams) public onlyOwner {
+        params = newParams;
+    }
+
+    function getParams() public view returns (RewardParams memory) {
+        return params;
+    }
+
+    function updateRewardsForAuctions(uint256 lastNounId) public {
+        uint256 currentlyAuctionedNounId = auctionHouse.auction().nounId;
+        require(lastNounId < currentlyAuctionedNounId, 'lastNounId must be settled');
```

Comment:
the two requirements in this function mean that if we even stop auctions, the last auction can't be processed. I don't mind it so much, but if we take this approach I think it's good to add some dev natspecs explaining why you're doing this vs adding an OR condition allowing it to be the current noun on auction if it's settled = true.

wdyt?

Reply:
agree. haven't given this edge case much thought.